### PR TITLE
gamescope: 3.11.49 -> 3.11.52-beta6

### PR DESCRIPTION
**Pull request by Scrumplex**
NOTE: This PR will probably be updated a few times until the eventual release of gamescope 3.11.52

- openvr: init at 1.23.8:
  - SDK for interacting with SteamVR, [Project page](https://github.com/ValveSoftware/openvr)
  - Patches are the same [as used by Arch Linux](https://github.com/archlinux/svntogit-community/blob/b7924c88fae0fe2026b6cf91f47db4d542a35564/trunk/PKGBUILD)
  - Required by gamescope for running as an OpenVR overlay
- libliftoff: 0.3.0 -> 0.4.0
  - [Changelog](https://gitlab.freedesktop.org/emersion/libliftoff/-/tags/v0.4.0), [Project page](https://gitlab.freedesktop.org/emersion/libliftoff)
  - Bump required by gamescope
- gamescope: 3.11.49 -> 3.11.52-beta6
  - [Project page](https://github.com/Plagman/gamescope)
  - Patches are here to fix the build in its current state. They are probably going to be unneeded for the stable release of 3.11.52

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Scrumplex** [comment]
As an aside: I would be interested in co-maintaining gamescope and maybe libliftoff, as I maintained packages for these on the Arch Linux User Repository, before moving to NixOS.

**Comment by PedroHLC** [comment]
@Scrumplex note that I have a few of the same in https://github.com/NixOS/nixpkgs/pull/215343 if you need something, kindly let me know.

**Comment by PedroHLC** [comment]
Result of `nixpkgs-review pr 219548` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>bottles</li>
    <li>bottles-unwrapped</li>
    <li>gamescope</li>
    <li>libliftoff</li>
    <li>openvr</li>
    <li>vkroots</li>
  </ul>
</details>

**Review by PedroHLC — APPROVED** [review]

**Comment by Scrumplex** [comment]
> @Scrumplex note that I have a few of the same in #215343 if you need something, kindly let me know.

:+1: 


> UPDATE: Also, Joshua said he does not want us to put vkroots as global package, since he'll never add a stable API to it, he believes it's better for every package to target a specific commit in the repo. But like you I did it anyway.

I just added vkroots inline in gamescope to address this. Especially as it's a header-only library it shouldn't really matter too much.

> UPDATE2: Put us both as co-maintainers of everything here 

done :D

> UPDATE3: Since you're adding the patches manually to a release tag, I could totally rebase my POC in this branch and stop bumping gamescope to untagged refs.

sure :D

File: pkgs/applications/window-managers/gamescope/default.nix [hunk 1]
```diff
@@ -18,15 +19,25 @@
 , seatd
 , xwayland
 , glslang
+, hwdata
+, openvr
 , stb
-, wlroots
+, wlroots_0_16
```
Comment (superherointj) [inline pkgs/applications/window-managers/gamescope/default.nix:25]:
```suggestion
, wlroots_0_16
```
Pinned version are usually inherited from top-level.

**Comment by superherointj** [comment]
If patches are the same would be good to `fetchpatch` from there.

File: pkgs/applications/window-managers/gamescope/default.nix [hunk 2]
```diff
@@ -73,8 +86,20 @@ stdenv.mkDerivation {
...
+    vkroots
+    libdisplay-info
   ];
 
+  postUnpack = ''
+    rm -rf source/subprojects/vkroots
+    mkdir source/subprojects/vkroots
+    ln -s ${vkroots}/* source/subprojects/vkroots
+    chmod -R +r+w source/subprojects/vkroots
+    chown -R $USER: source/subprojects/vkroots
```
Comment (superherointj) [inline pkgs/applications/window-managers/gamescope/default.nix:100]:
```suggestion
    ln -s ${vkroots}/* source/subprojects/vkroots
    chmod -R +r+w source/subprojects/vkroots
    chown -R $USER: source/subprojects/vkroots
```
Why not copy?

Comment (Scrumplex) [reply pkgs/applications/window-managers/gamescope/default.nix]: I just linked the whole directory now. Linking is way faster than copying

File: pkgs/applications/window-managers/gamescope/default.nix [hunk 3]
```diff
@@ -35,10 +46,10 @@ stdenv.mkDerivation {
     owner = "Plagman";
     repo = "gamescope";
     rev = "refs/tags/${version}";
-    hash = "sha256-GRq/b013wFRHzFz2YCulJRtcwzX/dhJKd8dkATSLug0=";
+    hash = "sha256-2gn6VQfmwwl86mmnRh+J1uxSIpA5x/Papq578seJ3n8=";
   };
 
-  patches = [ ./use-pkgconfig.patch ];
+  patches = [ ./use-pkgconfig.patch ./fix-openvr-dep-name.patch ./fix-openvr-include.patch ./bump-libdisplay-info-version.patch ./fix-libdisplay-info-dep-name.patch ];
```
[the remark this answers is not in the document]
Comment (Scrumplex) [reply pkgs/applications/window-managers/gamescope/default.nix]: done

File: pkgs/applications/window-managers/gamescope/default.nix (hunk 1 above)
Comment (Scrumplex) [reply pkgs/applications/window-managers/gamescope/default.nix]: done

File: pkgs/development/libraries/libliftoff/default.nix [hunk 4]
```diff
@@ -5,14 +5,14 @@
 
 stdenv.mkDerivation rec {
   pname = "libliftoff";
-  version = "0.3.0";
+  version = "0.4.0";
```
Comment (PedroHLC) [inline pkgs/development/libraries/libliftoff/default.nix:8]: You can safely bump this to 0.4.1 (I've already tested it)

**Review by PedroHLC — COMMENTED** [review]
I'll test this in my system, just a sec.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 219548` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>bottles</li>
    <li>bottles-unwrapped</li>
    <li>gamescope</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>libliftoff</li>
    <li>openvr</li>
  </ul>
</details>

**Comment by PedroHLC** [comment]
```console
/nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40/bin/ld: /nix/store/8s3ancq1nkbv1ir7n17yxhpdzvsak6i2-openvr-1.23.8/lib/libopenvr_api.so: undefined reference to `Json::Value::operator=(Json::Value)'
/nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40/bin/ld: /nix/store/8s3ancq1nkbv1ir7n17yxhpdzvsak6i2-openvr-1.23.8/lib/libopenvr_api.so: undefined reference to `Json::Value::operator!() const'
```

File: pkgs/development/libraries/openvr/default.nix [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  buildInputs = [ jsoncpp libGL ];
+
+  cmakeFlags = [ "-DUSE_SYSTEM_JSONCPP=ON" "-DBUILD_SHARED=1" ];
+
+  meta = with lib;{
+    homepage = "https://github.com/ValveSoftware/openvr";
+    description = "An API and runtime that allows access to VR hardware from multiple vendors without requiring that applications have specific knowledge of the hardware they are targeting";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ pedrohlc Scrumplex ];
+    platforms = with platforms; unix;
```
Comment (superherointj) [inline pkgs/development/libraries/openvr/default.nix:46]:
```suggestion
    platforms = platforms.unix;
```

**Comment by Scrumplex** [comment]
Looks like `fetchpatch` actually breaks patches (from GitHub) that rename files. The OpenVR patch to support the system installation of jsoncpp, also included headers of the in-tree jsoncpp, which is why there was this mismatch.

Am I missing a specific option of `fetchpatch` here?

**Comment by superherointj** [comment]
> Looks like `fetchpatch` actually breaks patches (from GitHub) that rename files. The OpenVR patch to support the system installation of jsoncpp, also included headers of the in-tree jsoncpp, which is why there was this mismatch.
> 
> Am I missing a specific option of `fetchpatch` here?

Try to reference the commit not the PR. We don't reference the PR, only commits. Because PR can change.

**Comment by Scrumplex** [comment]
Seems like this is a known issue (#32084). I also tried using `fetchpatch2` and also tried to pull in the commits separately, but it doesn't work.

This is what I got when I tried fetchpatch2:
```
error: Normalized patch '/build/patch' is empty (while the fetched file was not)!
Did you maybe fetch a HTML representation of a patch instead of a raw patch?
Fetched file was:
From a7a6995d1d6f9ee1a17a3741661156f9706c40ce Mon Sep 17 00:00:00 2001
From: Lubosz Sarnecki <lubosz.sarnecki@collabora.com>
Date: Thu, 15 Aug 2019 15:05:34 +0200
Subject: [PATCH] thirdparty: Move jsoncpp to thridparty directory.

Even though the previous patch in this patch set enabled the build with
a system wide jsoncpp and worked for me, it still left the internal
jsoncpp includes in the include path, since the includes were placed
in `CMAKE_CURRENT_SOURCE_DIR`. This could cause problems on other systems
when trying to build with a system wide jsoncpp.

In order to remove the internal json.h from the include path,
I moved all jsoncpp files into a thridparty directory amd include
it in the case of `USE_SYSTEM_JSONCPP` not being set.
---
 {src => thirdparty/jsoncpp}/json/json-forwards.h | 0
 {src => thirdparty/jsoncpp}/json/json.h          | 0
 {src => thirdparty/jsoncpp}/jsoncpp.cpp          | 0
 3 files changed, 0 insertions(+), 0 deletions(-)
 rename {src => thirdparty/jsoncpp}/json/json-forwards.h (100%)
 rename {src => thirdparty/jsoncpp}/json/json.h (100%)
 rename {src => thirdparty/jsoncpp}/jsoncpp.cpp (100%)

diff --git a/src/json/json-forwards.h b/thirdparty/jsoncpp/json/json-forwards.h
similarity index 100%
rename from src/json/json-forwards.h
rename to thirdparty/jsoncpp/json/json-forwards.h
diff --git a/src/json/json.h b/thirdparty/jsoncpp/json/json.h
similarity index 100%
rename from src/json/json.h
rename to thirdparty/jsoncpp/json/json.h
diff --git a/src/jsoncpp.cpp b/thirdparty/jsoncpp/jsoncpp.cpp
similarity index 100%
rename from src/jsoncpp.cpp
rename to thirdparty/jsoncpp/jsoncpp.cpp
```

**Comment by superherointj** [comment]
> Seems like this is a known issue (#32084). I also tried using `fetchpatch2` and also tried to pull in the commits separately, but it doesn't work.

Feel free to include the patch file then.

**Comment by Scrumplex** [comment]
> Feel free to include the patch file then.

Well in this case, we only really need to deal with a directory being renamed, so I am just going to add a postUnpack stage.

**Comment by superherointj** [comment]
> 

Great!

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 219548` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>bottles</li>
    <li>bottles-unwrapped</li>
    <li>gamescope</li>
    <li>libliftoff</li>
    <li>openvr</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Review by PedroHLC — APPROVED** [review]
Tested it in a live system, gamescope was able to handle CDPR launcher and the Witcher 3 as expected.

**Review by NguyenCuong17012021 — COMMENTED** [review]
Tess

**Comment by superherointj** [comment]
> Tess

What do you mean by `Tess`?

**Pull request merged**